### PR TITLE
🎁 `RewardTrainer` refactor

### LINKED ISSUE #3780: [Feature Request] Add support for padding-free reward modeling training in TRL
### Feature request

I would like to request support for padding-free reward modeling training in the HuggingFace TRL library, specifically in the RewardTrainer.

Currently, TRL's SFT trainer already supports padding-free training via position_ids, allowing for efficient fine-tuning on variable-length sequences. However, RewardTrainer still relies on padded inputs.

It would be very helpful if RewardTrainer could support a similar mechanism—by allowing user-supplied position_ids and attention_mask—to enable training without padding overhead.

### Motivation

Reward modeling datasets typically involve prompt + response pairs of varying lengths. Padding these to a uniform length introduces significant memory and compute inefficiencies, especially for long sequences and large batch sizes.

TRL’s SFT trainer has addressed this via padding-free training using custom position_ids. Bringing this same capability to RewardTrainer would streamline fine-tuning pipelines and reduce resource waste. This is particularly important in large-scale setups or when using FlashAttention-based architectures.

### Your contribution

Yes, I’d be happy to help by contributing a PR.

However, I believe this feature may require changes not only to TRL but also to HuggingFace Transformers—specifically in how forward() is implemented for SequenceClassification models. Many of these currently return only pooled_logits, which makes it difficult to apply position-aware token-level masking or sequence-level selection needed for padding-free reward modeling.

I’m happy to coordinate with the Transformers team or align with ongoing architectural changes to ensure compatibility.

### PR DESCRIPTION
**Pull request by qgallouedec**
This PR refactors `RewardTrainer`

closes #3780 #3101 #4104 #2633 #2758

- Aims for an implementation that better aligns with `SFTTrainer`, improving long-term maintainability  
- Streamlines usage for end-users (see readme diff)
- Better test coverage: expanded from 8 to 29 cases
- No regressions or breaking changes expected
- Better documentation: expanded from 94 to 245 lines
- Now works seamlessly with the CLI: `trl reward` ...
- Now includes its own model card
- Now supports activation offloading
- While caution is always good, this trainer isn’t heavily used, so the risk of major disruption is minimal (and we’re 99% confident nothing’s broken)
- The goal is to have the same refactor for other trainer (KTO, DPO, PRM, ...)

blue: before refactor
red: after refactor

Screenshot 2025-09-22 at 11 33 10 AM

to have the same base model:
```python
from transformers import AutoModelForSequenceClassification, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")
model = AutoModelForSequenceClassification.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct", num_labels=1)
tokenizer.save_pretrained("Qwen2.5-0.5B-Reward-Base")
model.saved_pretrained("Qwen2.5-0.5B-Reward-Base")
```

```python
# Code user before refactoring
from trl import RewardConfig, RewardTrainer
from datasets import load_dataset
from transformers import AutoModelForSequenceClassification, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("Qwen2.5-0.5B-Reward-Base")
model = AutoModelForSequenceClassification.from_pretrained("Qwen2.5-0.5B-Reward-Base", num_labels=1)
model.config.pad_token_id = tokenizer.pad_token_id

dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train[:2000]")

training_args = RewardConfig(output_dir="Qwen2.5-0.5B-Reward", learning_rate=0.0001, logging_steps=10)
trainer = RewardTrainer(
    args=training_args,
    model=model,
    processing_class=tokenizer,
    train_dataset=dataset,
)
trainer.train() # + needed to implement accuracy logging


# Code user after ractroing
from trl import RewardTrainer, RewardConfig
from datasets import load_dataset

dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train[:2000]")

trainer = RewardTrainer(
    model="Qwen2.5-0.5B-Reward-Base",
    train_dataset=dataset,
    args=RewardConfig(output_dir="Qwen2.5-0.5B-Reward", learning_rate=0.0001, logging_steps=10),
)
trainer.train()

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -72,40 +72,6 @@
...
-        return [remove_none_values(value) if isinstance(value, (dict, list)) else value for value in example]
-    elif isinstance(example, Mapping):
-        return {
-            key: remove_none_values(value) if isinstance(value, (dict, list)) else value
-            for key, value in example.items()
-            if value is not None
-        }
-    else:
-        raise TypeError("Input must be a list or a dictionary.")
-
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py:110]: moved to utils

File: trl/trainer/utils.py [hunk 2]
```diff
@@ -1251,9 +1254,10 @@ def generate_model_card(
         model_name=model_name,
         tags=["generated_from_trainer", *tags],
     )
+    template_file = template_file or "lm_model_card.md"
```
Comment (qgallouedec) [inline trl/trainer/utils.py:1335]: add a different model template file for reward, as the model is not a CausalLM, it's a SequenceClassification, so the code example was not accurate

File: trl/trainer/sft_config.py [hunk 3]
```diff
@@ -58,7 +58,7 @@ class SFTConfig(TrainingArguments):
         eos_token (`str`, *optional*):
             Token used to indicate the end of a turn or sequence. If `None`, it defaults to
             `processing_class.eos_token`.
-        pad_token (`int`, *optional*):
+        pad_token (`str`, *optional*):
```
Comment (qgallouedec) [inline trl/trainer/sft_config.py:61]: unrelated small fix

File: trl/trainer/reward_config.py [hunk 4]
```diff
@@ -70,37 +99,67 @@ class may differ from those in [`~transformers.TrainingArguments`].
...
+
+    # Parameters that control the training
     center_rewards_coefficient: Optional[float] = field(
         default=None,
         metadata={
             "help": "Coefficient to incentivize the reward model to output mean-zero rewards (proposed by "
             "https://huggingface.co/papers/2312.09244, Eq. 2). Recommended value: `0.01`."
         },
     )
-    remove_unused_columns: bool = field(
```
Comment (qgallouedec) [inline trl/trainer/reward_config.py:95]: we don't need `remove_unused_columns` to be false anymore, so we just rely on the super() default value.

File: trl/trainer/reward_config.py [hunk 5]
```diff
@@ -32,22 +32,51 @@ class may differ from those in [`~transformers.TrainingArguments`].
...
-            dataset is pretokenized.
+        activation_offloading (`bool`, *optional*, defaults to `False`):
+            Whether to offload the activations to the CPU.
     """
 
+    _VALID_DICT_FIELDS = TrainingArguments._VALID_DICT_FIELDS + ["model_init_kwargs"]
+
     # Parameters whose default values are overridden from TrainingArguments
+    learning_rate: float = field(
+        default=1e-4,
```
Comment (qgallouedec) [inline trl/trainer/reward_config.py:79]: better default for the learning rate. Found https://arxiv.org/pdf/2312.09244

File: trl/templates/rm_model_card.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+## Quick start
+
+```python
+from transformers import pipeline
+
+text = "The capital of France is Paris."
+rewarder = pipeline(model="{{ hub_model_id }}", device="cuda")
+output = rewarder(text)[0]
+print(output["score"])
```
Comment (qgallouedec) [inline trl/templates/rm_model_card.md:18]: The example is the only change compared to `lm_model_card.md`

File: tests/test_reward_trainer.py
Comment (qgallouedec) [inline tests/test_reward_trainer.py:1]: most of these tests are taken from tests/test_sft_trainer.

File: scripts/generate_tiny_models.py [hunk 7]
```diff
@@ -234,21 +236,36 @@ def init_weights_tiny_model(model):
...
+    config = AutoConfig.from_pretrained(model_id)
+    config.hidden_size = 16
+    config.num_attention_heads = 4
+    config.num_key_value_heads = 2
+    config.num_hidden_layers = 2
+    config.num_labels = 1
+    if model_id in ("Qwen/Qwen2.5-32B-Instruct", "Qwen/Qwen3-4B"):
+        config.layer_types = config.layer_types[:2]
+    model = model_class(config)
+    push_to_hub(model, tokenizer, "tiny", suffix)
```
Comment (qgallouedec) [inline scripts/generate_tiny_models.py:260]: this is a better way to create tiny models: just modify the attributes that determines the model size, and keep the other attributes.

File: scripts/generate_tiny_models.py [hunk 8]
```diff
@@ -234,21 +236,36 @@ def init_weights_tiny_model(model):
 push_to_hub(model, tokenizer, "small")
 
 # Reward models
-for model_id, config_class, model_class, suffix in [
-    ("meta-llama/Llama-3.2-1B-Instruct", LlamaConfig, LlamaForSequenceClassification, "3.2"),
-    ("Qwen/Qwen2.5-32B-Instruct", Qwen2Config, Qwen2ForSequenceClassification, "2.5"),
-    ("Qwen/Qwen3-4B", Qwen3Config, Qwen3ForSequenceClassification, None),
+for model_id, model_class, suffix in [
+    ("EleutherAI/pythia-14m", GPTNeoXForSequenceClassification, None),
```
Comment (qgallouedec) [inline scripts/generate_tiny_models.py:240]: new tiny model: useful because this one has no chat template (the other models have)

File: scripts/generate_tiny_models.py [hunk 9]
```diff
@@ -234,21 +236,36 @@ def init_weights_tiny_model(model):
...
+    config.num_hidden_layers = 2
+    config.num_labels = 1
+    if model_id in ("Qwen/Qwen2.5-32B-Instruct", "Qwen/Qwen3-4B"):
+        config.layer_types = config.layer_types[:2]
+    model = model_class(config)
+    push_to_hub(model, tokenizer, "tiny", suffix)
+
+# MoE Reward models
+for model_id, model_class, suffix in [
+    ("Qwen/Qwen3-30B-A3B", Qwen3MoeForSequenceClassification, None),
```
Comment (qgallouedec) [inline scripts/generate_tiny_models.py:264]: we now test MoE reward models!

File: docs/source/reward_trainer.md
Comment (qgallouedec) [inline docs/source/reward_trainer.md:1]: I recommend directly checking the rendered version, I re-wrote everything, the diff doesn't really helps here

File: README.md [hunk 10]
```diff
@@ -136,23 +136,13 @@ trainer.train()
...
 
 dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train")
 
-training_args = RewardConfig(output_dir="Qwen2.5-0.5B-Reward", per_device_train_batch_size=2)
 trainer = RewardTrainer(
-    args=training_args,
-    model=model,
-    processing_class=tokenizer,
+    model="Qwen/Qwen2.5-0.5B-Instruct",
     train_dataset=dataset,
```
Comment (qgallouedec) [inline README.md:146]: so much easier 🤤

File: trl/trainer/reward_trainer.py [hunk 11]
```diff
@@ -12,217 +12,398 @@
...
-            Model to be trained, preferably an [`~transformers.AutoModelForSequenceClassification`].
+        model (`Union[str, PreTrainedModel]`):
+            Model to be trained. Can be either:
+
+            - A string, being the *model id* of a pretrained model hosted inside a model repo on huggingface.co, or a
+              path to a *directory* containing model weights saved using
+              [`~transformers.PreTrainedModel.save_pretrained`], e.g., `'./my_model_directory/'`. The model is loaded
+              using `AutoModelForSequenceClassification.from_pretrained` with the keyword arguments in
+              `args.model_init_kwargs`.
+            - A [`~transformers.PreTrainedModel`] object.
```
Comment (qgallouedec) [inline trl/trainer/reward_trainer.py]:
```suggestion
            - A sequence classification [`~transformers.PreTrainedModel`] object.
```

File: docs/source/reward_trainer.md [hunk 12]
```diff
@@ -2,84 +2,231 @@
...
+
+The reward model  \\( r_\theta(x, y) \\) is trained to assign higher scores to preferred responses  \\( y^+ \\) over non-preferred ones  \\( y^- \\). The loss is then defined as the negative log-likelihood of the observed preferences:
+
+$$
+\mathcal{L}(\theta) = - \mathbb{E}_{(x,y^+,y^-) \sim \mathcal{D}} \left[ \log \sigma(r_\theta(x, y^+) - r_\theta(x, y^-)) \right].
+$$
+
+<Tip>
+
+The Bradley-Terry model is underdetermined, meaning that adding a constant to all rewards does not change the preference probabilities. To address this, [Helping or Herding? Reward Model Ensembles Mitigate but do not Eliminate Reward Hacking](https://huggingface.co/papers/2312.09244) proposes adding an auxiliary loss term that encourages the rewards to be centered around zero. This is controlled by the `center_rewards_coefficient` parameter in the [`RewardConfig`]. The recomended value is `1e-2`.
```
Comment (kashif) [inline docs/source/reward_trainer.md]:
```suggestion
The Bradley-Terry model is underdetermined, meaning that adding a constant to all rewards does not change the preference probabilities. To address this, [Helping or Herding? Reward Model Ensembles Mitigate but do not Eliminate Reward Hacking](https://huggingface.co/papers/2312.09244) proposes adding an auxiliary loss term that encourages the rewards to be centered around zero. This is controlled by the `center_rewards_coefficient` parameter in the [`RewardConfig`]. The recommended value is `1e-2`.
```
